polyply/src/gen_seq.py

```diff
@@ -0,0 +1,189 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+import json
+import networkx as nx
+from networkx.readwrite import json_graph
+from .load_library import load_library
+from .generate_templates import find_atoms
+
+def _macro_to_graph(resname, branching_f, n_levels):
+    """
+    Generate a tree graph of branching factor `branching_f`
+    and number of generations as `n_levels` - 1. Each node
+    gets the attribute resname.
+
+    Parameters
+    ----------
+    resname:  str
+    branching_f: int
+    n_levels: int
+
+    Returns
+    -------
+    `:class:networkx.Graph`
+    """
+    graph = nx.balanced_tree(r=branching_f, h=n_levels-1)
+    resnames = {idx: resname for idx in graph.nodes}
+    nx.set_node_attributes(graph, resnames, "resname")
+    return graph
+
+def _add_edges(graph, edges, idx, jdx):
+    """
+    Add edges to a graph using the edge format
+    'node_1,node_2', where the node keys refere
+    to the node relative to the nodes with the same
+    attribute `seqid`. The sequence ids are provided by
+    `idx` and `jdx`.
+
+    Parameters
+    ----------
+    graph: `:class:networkx.Graph`
+    edges:  list[str]
+        str has the format `node_key,node_key`
+    idx: int
+    jdx: int
+
+    Returns
+    --------
+    `:class:networkx.Graph`
+    """
+    for edge in edges.split(","):
+        node_idx, node_jdx = edge.split("-")
+        idx_nodes = find_atoms(graph, "seqid", idx)
+        jdx_nodes = find_atoms(graph, "seqid", jdx)
+        graph.add_edge(idx_nodes[int(node_idx)], jdx_nodes[int(node_jdx)])
+    return graph
+
+def _random_macro_to_graph(n_blocks, residues):
+    """
+    Generate a linear graph with `n_blocks` nodes,
+    and set a node attribute according to the probabilites
+    defined in `residues`. Residues is a list of strings
+    in the form of residue_1-precentage,residue_2-percantage.
+    The percantages needs to add up to 1.
+
+    Parameters
+    ----------
+    n_blocks: int
+    residues: list[str]
+
+    Returns:
+    --------
+    `:class:networkx.Graph`
+    """
+    macro_graph = _macro_to_graph("dum", 1, int(n_blocks))
+    nodes = list(macro_graph.nodes)
+    len_graph = len(nodes)
+    for res_prob in residues.split(','):
+        resname, percentage = res_prob.split('-')
+        # this can probably be done smater
+        nnodes = int(float(percentage) * len_graph)
+        chosen_nodes = random.sample(nodes, k=nnodes)
+        res_attr = {node: resname for node in chosen_nodes}
```

Comment:
This is not truly random, is it? Or at least not the way I would expect it. I would expect every residue to have a p% chance of being A, and p2% chance of being B. This is not the same as saying that p% of all residues will be A!

I'd advocate the use of random.choices with the percentages as weights.